File: computer-sience/algorithms/sorting/sorting.cpp
#include "sorting.h"
#include <vector>


template <typename T>
void swap(std::size_t i, std::size_t j, std::vector<T>& arr) 
{
    std::vector<T> temp;
    temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}


template <typename T>
void SelectionSort(std::vector<T> arr) 
{
    auto n = arr.size();
    for (auto i=0;i<n;i++) 
    {
        auto MinIdx = i;
        for (auto j=0;j<n-1;j++) 
        {
            if (arr[MinIdx] > arr[j]) 
            {
                MinIdx = j;
            }
        }
        swap(i, MinIdx, arr);
    }
}


template <typename T>
void BubbleSort(std::vector<T> arr)
{
    auto n = arr.size();
    for (auto i=0;i<n-1;i++) 
    {
        for (auto j=i;j<n-1;j++) 
        {
            if (arr[j] > arr[j+1] )
            {
                swap(i, j, arr);
            }
        }
        
    }
}

File: computer-sience/algorithms/sorting/sorting.go
package sorting

import "cmp"

func selectionSort[T cmp.Ordered](arr []T) {
	n := len(arr)

	for i := 0; i < n; i++ {
		minIdx := i
		for j := i + 1; j < n; j++ {
			if arr[minIdx] > arr[j] {
				minIdx = j
			}
		}
		arr[i], arr[minIdx] = arr[minIdx], arr[i]
	}
}

func bubbleSort[T cmp.Ordered](arr []T) {
	n := len(arr)
	for i := 1; i < n; i++ {
		for j := 0; j < n-1; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}
}

func insertionSort[T cmp.Ordered](arr []T) {
	n := len(arr)
	for i := 1; i < n; i++ {
		p := i - 1
		for p > 0 && arr[p+1] < arr[p] {
			arr[p], arr[p+1] = arr[p+1], arr[p]
		}
	}
}
